### PR TITLE
Generalise mapping to work on getters and reviews

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by arybczak** [comment]
Is that pretty much a read only part of `mapping` from `Optics.Iso`?

**Comment by arybczak** [comment]
Perhaps it's worth making `mapping` a part of an optic-kind specific type class then?

**Comment by phadej** [comment]
Hmm, good point.

```
mapping :: Iso -> Iso
mapping :: Lens -> Getter
mapping :: Getter -> Getter
mapping :: Prism -> Review
mapping :: Review -> Review
```

so you could `review (mapping _Just) [1,2,3]` e.g.

Does that sound right, @arybczak ?

**Comment by arybczak** [comment]
Yes, I think so :)

**Comment by phadej** [comment]
Pushed a generalised version

File: optics/optics.cabal [hunk 1]
```diff
@@ -83,6 +83,7 @@ library
                     , Optics.Empty
                     , Optics.Indexed
                     , Optics.Label
+                    , Optics.Mapping
```
Comment (arybczak) [inline optics/optics.cabal]: This needs to be re-exported in `Optics` module.

File: optics-core/src/Optics/Iso.hs [hunk 2]
```diff
@@ -150,13 +144,7 @@ under k = withIso k $ \sa bt ts -> sa . ts . bt
...
 -- Isomorphisms
 
--- | This can be used to lift any 'Iso' into an arbitrary 'Functor'.
-mapping
-  :: (Functor f, Functor g)
-  => Iso    s     t     a     b
-  -> Iso (f s) (g t) (f a) (g b)
-mapping k = withIso k $ \sa bt -> iso (fmap sa) (fmap bt)
-{-# INLINE mapping #-}
+
```
Comment (arybczak) [inline optics-core/src/Optics/Iso.hs]: Extra newline

File: optics-core/src/Optics/Iso.hs [hunk 3]
```diff
@@ -74,13 +74,7 @@ module Optics.Iso
...
   , under
 
-  -- * Combinators
-  -- | The 'Optics.Re.re' combinator can be used to reverse an 'Iso':
-  --
-  -- @
-  -- 'Optics.Re.re' :: 'Iso' s t a b -> 'Iso' b a t s
-  -- @
-  , mapping
+
```
Comment (arybczak) [inline optics-core/src/Optics/Iso.hs]: Extra newline

File: optics-core/src/Optics/Mapping.hs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+-- >>> import Optics.Core
+-- >>> import Optics.Operators
+
+-- | Class for optics supporting 'mapping' through a 'Functor'.
+class MappingOptic k f g s t a b where
+  -- | Type family that maps an optic to the optic kind produced by
+  -- 'mapping' using it.
+  type MappedOptic k
+
+  -- | The 'Optics.Re.mapping' can be used to lift optic through a 'Functor'.
```
Comment (arybczak) [inline optics-core/src/Optics/Mapping.hs]: `Optics.Re.Mapping` :)

**Comment by phadej** [comment]
ping @adamgundry

File: optics-core/src/Optics/Iso.hs [hunk 5]
```diff
@@ -74,14 +74,6 @@ module Optics.Iso
...
   , under
 
-  -- * Combinators
-  -- | The 'Optics.Re.re' combinator can be used to reverse an 'Iso':
-  --
-  -- @
-  -- 'Optics.Re.re' :: 'Iso' s t a b -> 'Iso' b a t s
-  -- @
-  , mapping
-
```
Comment (adamgundry) [inline optics-core/src/Optics/Iso.hs]: We seem to have lost some documentation here. I'd suggest keeping the mention of `mapping` but in Haddock comments, like `re`.

Comment (phadej) [reply optics-core/src/Optics/Iso.hs]: Fixed. I was confused, it looked like the above is "forgotten to update" documentation for `mapping`.
Resurrected it so *Combinators* section doesn't export anything, so it isnt' so confusing:

Screenshot from 2020-02-10 14-46-20

**Pull request merged**